src/language/term/Typ.re

```diff
@@ -561,16 +576,34 @@ let rec match_synswitch = (t1: t, t2: t) => {
   };
 };
 
-let join_all = (~empty: t, ctx: Ctx.t, ts: list(t)): option(t) =>
+let join_all =
+    (~inconsistent=?, ~empty: t, ctx: Ctx.t, ts: list(t)): option(t) =>
   List.fold_left(
-    (acc, ty) => OptUtil.and_then(join(ctx, ty), acc),
+    (acc, ty) => OptUtil.and_then(join(~inconsistent?, ctx, ty), acc),
     Some(empty),
     ts,
   );
 
 let is_consistent = (ctx: Ctx.t, ty1: t, ty2: t): bool =>
   join(ctx, ty1, ty2) != None;
 
+/*
+    Computes the most precise type that is consistent with all input types.
+    Like a multi-way join, but if any ground types are inconsistent, replaces
+    that position with an "unknown" type rather than failing.
+    This operation is NOT associative — applying it pairwise may yield
+    a different result than applying it to the whole list at once.
+ */
+let consistent_join = (ctx: Ctx.t, tys: list(t)): t => {
+  join_all(
+    ~inconsistent=Unknown(Inconsistent) |> temp,
+    ~empty=Unknown(SynSwitch) |> temp,
+    ctx,
+    tys,
+  )
+  |> Option.value(~default=Unknown(SynSwitch) |> temp);
+};
```

Comment:
This could be used for all the places we're using join to synthesize a type from inconsistent internal errors.